exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java

```diff
@@ -465,9 +498,77 @@ private static Resource getAutoDetectedResourceType() {
     if (MetadataConfig.getInstanceId() != null) {
       return Resource.GceInstance;
     }
+
+    InputStream stream = null;
+    try {
+      stream = openStream(AWS_INSTANCE_IDENTITY_DOCUMENT_URI);
+      String awsIdentityDocument = slurp(new InputStreamReader(stream, Charsets.UTF_8));
+      awsEnvVarMap = parseAwsIdentityDocument(awsIdentityDocument);
+      return Resource.AwsEc2Instance;
+    } catch (IOException e) {
+      // Cannot connect to http://169.254.169.254/latest/dynamic/instance-identity/document.
+      // Not on an AWS EC2 instance.
+    } finally {
+      if (stream != null) {
+        try {
+          stream.close();
+        } catch (IOException e) {
+          // Do nothing.
+        }
+      }
+    }
+
     // default Resource type
     return Resource.Global;
   }
 
+  /** quick http client that allows no-dependency try at getting instance data. */
+  private static InputStream openStream(URI uri) throws IOException {
+    HttpURLConnection connection = HttpURLConnection.class.cast(uri.toURL().openConnection());
+    connection.setConnectTimeout(1000 * 2);
+    connection.setReadTimeout(1000 * 2);
+    connection.setAllowUserInteraction(false);
+    connection.setInstanceFollowRedirects(false);
+    return connection.getInputStream();
+  }
+
+  /** returns the {@code reader} as a string without closing it. */
+  private static String slurp(Reader reader) throws IOException {
+    StringBuilder to = new StringBuilder();
+    CharBuffer buf = CharBuffer.allocate(BUF_SIZE);
+    while (reader.read(buf) != -1) {
+      buf.flip();
+      to.append(buf);
+      buf.clear();
+    }
+    return to.toString();
+  }
+
+  // AWS Instance Identity Document is a JSON file.
+  // See docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html.
+  @VisibleForTesting
+  static Map<String, String> parseAwsIdentityDocument(String awsIdentityDocument) {
+    Map<String, String> map = Maps.newHashMap();
+    List<String> lines = LINE_BREAK_SPLITTER.splitToList(awsIdentityDocument);
+    for (String line : lines) {
+      List<String> keyValuePair = COLON_SPLITTER.splitToList(line);
+      if (keyValuePair.size() != 2) {
+        continue;
+      }
+      String key = keyValuePair.get(0).replaceAll("[\" ]", "");
+      String value = keyValuePair.get(1).replaceAll("[\" ,]", "");
+      map.put(key, value);
+    }
+    return map;
+  }
+
+  @javax.annotation.Nullable
+  private static String getValueFromAwsIdentityDocument(String key) {
```

Comment:
I would make this AWS related parsing/calling metadata a separate package class. I think this will make the code cleaner and easier to test and read.

Reply:
Done.